Public/Src/Cache/ContentStore/Distributed/NuCache/LocalLocationStore.cs

```diff
@@ -1380,6 +1410,47 @@ public Task<BoolResult> RegisterLocalLocationAsync(OperationContext context, IRe
             }
         }
 
+        private Task<BoolResult> ProactiveReplicationAsync(OperationContext context)
+        {
+            return context.PerformOperationAsync(
+                Tracer,
+                async () =>
+                {
+                    IEnumerable<ContentInfo> localContent = await _localContentStore.GetContentInfoAsync(context.Token);
+                    localContent = localContent.OrderByDescending(info => info.LastAccessTimeUtc);
+
+                    var proactiveReplicationBlock = new ActionBlock<ContentHash>(
+                        async hash => await _proactiveReplicationTaskFactory(context, hash),
+                        new ExecutionDataflowBlockOptions()
+                        {
+                            MaxDegreeOfParallelism = _configuration.ProactiveReplicationConcurrencyLimit
+                        });
+
+                    var contentCopied = 0;
+                    foreach (var content in localContent)
+                    {
+                        if (Database.TryGetEntry(context, content.ContentHash, out var entry))
+                        {
+                            if (entry.Locations.Count < _configuration.ProactiveCopyLocationsThreshold)
+                            {
+                                await proactiveReplicationBlock.SendAsync(content.ContentHash);
+
+                                contentCopied++;

```

Comment:
We need a special tracing here similar to one we have in garbage collection with some stats: how may items were replicated, how many successes and failures etc.
And we need special counters for each individual replication.